tests/test_addons.py

```diff
@@ -38,4 +51,119 @@ def test_static_url_absolute(self):
             '/foo'
         )
 
-    # TODO: Add tests for callbacks - specifically, `on_add()` and `on_delete()`
+
+class SetEnvironMiddleware(object):
+
+    def __init__(self, app, **kwargs):
+        self.app = app
+        self.kwargs = kwargs
+
+    def __call__(self, environ, start_response):
+        environ.update(self.kwargs)
+        return self.app(environ, start_response)
+
+
+class TestAddonAuth(OsfTestCase):
+
+    def setUp(self):
+        super(TestAddonAuth, self).setUp()
+        self.flask_app = SetEnvironMiddleware(self.app.app, REMOTE_ADDR='127.0.0.1')
+        self.test_app = webtest.TestApp(self.flask_app)
+        self.user = AuthUserFactory()
+        self.auth_obj = Auth(user=self.user)
+        self.node = ProjectFactory(creator=self.user)
+        self.session = Session(data={'auth_user_id': self.user._id})
+        self.session.save()
+        self.cookie = itsdangerous.Signer(settings.SECRET_KEY).sign(self.session._id)
+        self.configure_addon()
+
+    def configure_addon(self):
+        self.user.add_addon('github')
+        self.user_addon = self.user.get_addon('github')
+        self.oauth_settings = AddonGitHubOauthSettings(github_user_id='john')
+        self.oauth_settings.save()
+        self.user_addon.oauth_settings = self.oauth_settings
+        self.user_addon.oauth_access_token = 'secret'
+        self.user_addon.save()
+        self.node.add_addon('github', self.auth_obj)
+        self.node_addon = self.node.get_addon('github')
+        self.node_addon.user = 'john'
+        self.node_addon.repo = 'youre-my-best-friend'
+        self.node_addon.user_settings = self.user_addon
+        self.node_addon.save()
+
+    def build_url(self, **kwargs):
+        options = dict(
+            action='download',
+            cookie=self.cookie,
+            token='',
+            nid=self.node._id,
+            provider=self.node_addon.config.short_name,
+        )
+        options.update(kwargs)
+        return api_url_for('get_auth', **options)
+
+    def test_auth_download(self):
+        url = self.build_url()
+        res = self.test_app.get(url)
+        assert_equal(res.json['auth'], views.make_auth(self.user))
+        assert_equal(res.json['credentials'], self.node_addon.serialize_waterbutler_credentials())
+        assert_equal(res.json['settings'], self.node_addon.serialize_waterbutler_settings())
```

Comment:
@CenterForOpenScience/waterbutler-admin Is `serialize_waterbutler_credentials()` a necessary override for all storage addons? Should it be an abstract method?